homeassistant/components/onkyo/config_flow.py

```diff
@@ -0,0 +1,235 @@
+            self._abort_if_unique_id_configured()
+
+            try:
+                with async_timeout.timeout(CONNECT_TIMEOUT):
+                    await self._connection.connect()
+
+            except asyncio.TimeoutError:
+                return self.async_abort(reason="cannot_connect")
+
+            # Close the test connection as setup entry will create one.
+            self._connection.close()
+            return self.async_create_entry(
+                title=self._connection.name,
+                data={
+                    CONF_IDENTIFIER: self._connection.identifier,
+                    CONF_HOST: self._connection.host,
+                    CONF_NAME: self._connection.name,
+                },
+            )
+
+        return self.async_abort(reason="unknown")
+
+
+class OptionsFlowHandler(config_entries.OptionsFlow):
+    """Handle options for the Onkyo component."""
+
+    def __init__(self, config_entry: config_entries.ConfigEntry) -> None:
+        """Initialize options flow."""
+        self.config_entry: config_entries.ConfigEntry = config_entry
+        self._options: dict[str, str | dict[str, Any]] = {}
+
+    async def async_step_init(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle options flow initialized by the user."""
+        # Create a dictionary with all sources and set custom source names.
+        sources = (
+            {**DEFAULT_SOURCE_NAMES, **self.config_entry.options[CONF_SOURCES]}
+            if CONF_SOURCES in self.config_entry.options
+            else {**DEFAULT_SOURCE_NAMES, **DEFAULT_SOURCES}
+        )
+
+        if user_input is not None:
+            # Store the input to update the entry after the next step.
+            self._options = {
+                CONF_MAX_VOLUME: user_input[CONF_MAX_VOLUME],
+                CONF_SOURCES: {
+                    key: source_name
+                    for key, source_name in sources.items()
+                    if key in user_input[CONF_ENABLED_SOURCES]
+                },
+            }
+            return await self.async_step_source_names()
+
+        settings_schema = vol.Schema(
+            {
+                vol.Optional(
+                    CONF_MAX_VOLUME,
+                    default=self.config_entry.options.get(
+                        CONF_MAX_VOLUME, DEFAULT_MAX_VOLUME
+                    ),
+                ): vol.All(vol.Coerce(int), vol.Range(min=1)),
+                vol.Optional(
+                    CONF_ENABLED_SOURCES,
+                    default=list(
+                        self.config_entry.options.get(
+                            CONF_SOURCES, DEFAULT_SOURCES
+                        ).keys()
+                    ),
+                ): cv.multi_select(
+                    {
+                        key: f"{source_name} ({key})"
+                        for key, source_name in sources.items()
+                    }
+                ),
```

Comment:
These are device-level options? Not integration options?

Reply:
I didn't know there's a difference, but these are indeed device-level options and may be different per added device. Do I have to change anything?

Reply:
I think max volume and such are device setting that can be just entities?

Enabled sources, I'm not sure, do we need it even? Why not enable all by default?

Reply:
What do you mean can be just entities? Users can use max_volume to limit the max volume of the receiver if 100% is too loud. So I think they should be able to configure it as an option. I think @OnFreund can agree with me here, he also helped out on the first [PR](https://github.com/home-assistant/core/pull/33694)

In the current integration the sources can also be set so I don't want to drop the functionality. It's also more clear to only see the inputs of the receiver you use.

Reply:
> What do you mean can be just entities? Users can use max_volume to limit the max volume of the receiver if 100% is too loud.

If the receiver has a max/min volume setting, it can be represented as an entity (number entity). 

> In the current integration the sources can also be set so I don't want to drop the functionality.

How would that drop functionality? The sources are still available.

Reply:
I've been playing around with changing the maximum volume to a number entity, but I'm not sure about a couple of things. The maximum volume of a receiver is either 80 or 200, depending on the model, but there is no easy way to read that info from the receiver, so the user should set it per device. 

But if you never want to have your receiver at max volume, you can also set the maximum volume lower so that HA will not set the volume above the set level.

Should all receivers then have a maximum volume number entity that ranges from 0 to 200? But it would not make sense to make this value higher then 80 on a receiver which max volume is 80. But if we want the range to stop at the max volume of the receiver, we're gonna need an option in the device option flow to set the max of the max volume range. 

Maybe @OnFreund has an idea which works best, keeping the max volume an option as it is already, or changing it to a number entity that always ranges from 0 to 200, or also make an option to set the max range of the number entity

Reply:
> we're gonna need an option in the device option flow to set the max of the max volume range.

Home Assistant doesn't have a concept of device option flows.

Reply:
Sorry, I should use the correct naming, I mean an integration configuration option, just like the max volume is implemented now.

Reply:
>  integration configuration option

Option flows for configurations should not really be used for "device" configuration.

Reply:
Okay thank you, shall I then just create a number entity for each device, although it will probably be set once?